src/tactic/equiv_rw.lean

```diff
@@ -0,0 +1,260 @@
+
+## Implementation details
+
+The main internal function is `adapt_equiv t e`,
+which attempts to turn an expression `e : α ≃ β` into a new equivalence with left hand side `t`.
+As an example, with `t = option α`, it will generate `functor.map_equiv option e`.
+
+This is achieved by generating a new synthetic goal `%%t ≃ _`,
+and calling `solve_by_elim` with an appropriate set of congruence lemmas.
+To avoid having to specify the relevant congruence lemmas by hand,
+we mostly rely on `functor.map_equiv` and `bifunctor.map_equiv`
+(and, in a future PR, `equiv_functor.map_equiv`, see below),
+along with some structural congruence lemmas such as
+* `equiv.arrow_congr'`,
+* `equiv.subtype_equiv_of_subtype'`,
+* `equiv.sigma_congr_left'`, and
+* `equiv.Pi_congr_left'`.
+
+The main `equiv_rw` function, when operating on the goal, simply generates a new equivalence `e'`
+with left hand side matching the target, and calls `apply e'.inv_fun`.
+
+When operating on a hypothesis `x : α`, we introduce a new fact `h : x = e.symm (e x)`,
+revert this, and then attempt to `generalize`, replacing all occurrences of `e x` with a new constant `y`,
+before `intro`ing and `subst`ing `h`, and renaming `y` back to `x`.
+
+## Future improvements
+While `equiv_rw` will rewrite under type-level functors (e.g. `option` and `list`),
+many constructions are only functorial with respect to equivalences
+(essentially, because of both positive and negative occurrences of the type being rewritten).
+At the moment, we only handle constructions of this form (e.g. `unique` or `ring`) by
+including explicit `equiv.*_congr` lemmas in the list provided to `solve_by_elim`.
+
+A (near) future PR will introduce `equiv_functor`,
+a typeclass for functions which are functorial with respect to equivalences,
+and use these instances rather than hard-coding the `congr` lemmas here.
+
+`equiv_functor` is not included in the initial implementation of `equiv_rw`, simply because
+the best way to construct an instance of `equiv_functor has_add` is to use `equiv_rw`!
+In fact, in a slightly more distant future PR I anticipate that
+`derive equiv_functor` should work on many examples, and
+we can incrementally bootstrap the strength of `equiv_rw`.
+
+An ambitious project might be to add `equiv_rw!`,
+a tactic which, when failing to find appropriate `equiv_functor` instances,
+attempts to `derive` them on the spot.
+
+For now `equiv_rw` is entirely based on `equiv` in `Type`,
+but the framework can readily be generalised to also work with other types of equivalences,
+for example specific notations such as ring equivalence (`≃+*`),
+or general categorical isomorphisms (`≅`).
+
+This will allow us to transport across more general types of equivalences,
+but this will wait for another subsequent PR.
+-/
+
+namespace tactic
+
+/-- A hard-coded list of names of lemmas used for constructing congruence equivalences. -/
+
+-- TODO: This should not be a hard-coded list.
```

Comment:
I guess the idea is that once `equiv_functor` lands, this "hard-coded" list will not need to change at all as the library grows, because to extend the scope of `equiv_rw` you'll just need to create new instances of `functor`, `bifunctor`, or `equiv_functor` (and perhaps `equiv_bifunctor` if that becomes necessary).

Making the list into an attribute thus has two disadvantages:
1. it lets users add to it, when I'd really prefer they construct the functoriality instances!
2. it's harder to control the ordering, which is pretty important for correct behaviour.